### PR TITLE
[compiler-rt] Realtime Sanitizer: Introduce Realtime Sanitizer (RTSan) backend

### PR DESCRIPTION
**Pull request by cjappl**
Introducing the main runtime of realtime sanitizer. For more information, please see the [discourse thread](https://discourse.llvm.org/t/rfc-nolock-and-noalloc-attributes/76837). We have also put together a [reviewer support document](https://github.com/realtime-sanitizer/radsan/blob/doc/review-support/doc/review.md) to show what our intention is.

This review introduces the sanitizer backend. This includes:
* CMake build files (largely adapted from asan).
* Main RTSan architecture (the external API, thread local context, stack).
* Interceptors.
* Many unit tests.

Please see the [reviewer support document](https://github.com/realtime-sanitizer/radsan/blob/doc/review-support/doc/review.md) for what our next steps are. We are moving in lockstep with this PR #84983 for the codegen coming up next.

Note to reviewers: If you see support documentation mention "RADSan", this was the "old acronym" for the realtime sanitizer, they refer to the same thing. If you see it let us know and we can correct it (especially in the llvm codebase)

### CODE REVIEW

File: clang/lib/Driver/ToolChains/CommonArgs.cpp
Comment (cjappl) [inline clang/lib/Driver/ToolChains/CommonArgs.cpp:1387]: One question we have here: What is the SharedRuntimes vs StaticRuntimes? Why would we need to support both? This is one change we "eyeballed" based on the other sanitizers.

File: compiler-rt/cmake/Modules/AllSupportedArchDefs.cmake
Comment (cjappl) [inline compiler-rt/cmake/Modules/AllSupportedArchDefs.cmake:37]: This was copied from ASAN, is this acceptable? We don't have all architectures to test, although we think that our interceptors are simple enough to work on any machine (no platform specific assembly in our code, for example)

**Comment by cjappl** [comment]
Pinging possibly interested parties for review. 

@yln @vitalybuka @Sirraide @AaronBallman @dougsonos @davidtrevelyan 

Please feel free to add more, we don't know who all may be interested. Thanks in advance

File: clang/lib/Driver/ToolChains/CommonArgs.cpp
Comment (MaskRay) [reply clang/lib/Driver/ToolChains/CommonArgs.cpp:1387]: I have some notes at https://maskray.me/blog/2023-01-08-all-about-sanitizer-interceptors#static-runtime

Nit: we omit braces in this single-line statement case https://llvm.org/docs/CodingStandards.html#don-t-use-braces-on-simple-single-statement-bodies-of-if-else-loop-statements

File: compiler-rt/lib/radsan/radsan.h
Comment (MaskRay) [inline compiler-rt/lib/radsan/radsan.h:29]: Use `__*` names to avoid conflicts with user code. Use `//`.

If there are user APIs, define them at `include/sanitizer/*_interface.h`

File: compiler-rt/test/radsan/lit.cfg.py
Comment (cjappl) [reply compiler-rt/test/radsan/lit.cfg.py:9]: Yes, the lit tests may be placed directly in `compiler-rt/test/radsan/` and they run as intended. Just checked on our branch that has them implemented.

File: compiler-rt/lib/radsan/radsan_context.cpp
Comment (cjappl) [reply compiler-rt/lib/radsan/radsan_context.cpp:27]: Great thanks!! Before I go through and edit these, `snake_case` is preferred for variables in all contexts?

Really appreciate the review. Pushing up most of your requested changes now.

Comment (MaskRay) [reply compiler-rt/lib/radsan/radsan_context.cpp:27]: Yes, `variable_name` and `FunctionName`.. Sorry if this change may take a while.
asan/hwasan/lsan are good examples to follow.

Comment (cjappl) [reply compiler-rt/lib/radsan/radsan_context.cpp:27]: No apology necessary! Happy to do this right the first time.

Does this also apply to lambdas?

```
  auto Func = [&vec]() { vec.push_back(0.4f); };
```

Should this become:
```
  auto func = [&vec]() { vec.push_back(0.4f); };
```

Or remain "function case"?

**Comment by MaskRay** [comment]
https://discourse.llvm.org/t/rfc-nolock-and-noalloc-attributes/76837/75 contains a discussion about the subproject name.
Do you have opinions? @llvm/pr-subscribers-sanitizer @compnerd @petrhosek  @isanbard

**Comment by cjappl** [comment]
Weekly ping of reviewers

@yln @vitalybuka @Sirraide @AaronBallman @zygoloid @compnerd @petrhosek @isanbard

Looking for weighing in on any of the code as it stands now, or the naming question posed by MaskRay [here](https://discourse.llvm.org/t/rfc-nolock-and-noalloc-attributes/76837/75), so far there seems to be support for rtsan over radsan :)

**Comment by cjappl** [comment]
Ping

@yln @vitalybuka @Sirraide @AaronBallman @zygoloid @compnerd @petrhosek @isanbard @MaskRay

**Comment by dougsonos** [comment]
> Looking for weighing in on any of the code as it stands now, or the naming question posed by MaskRay [here](https://discourse.llvm.org/t/rfc-nolock-and-noalloc-attributes/76837/75), so far there seems to be support for rtsan over radsan :)

+1 for rtsan

File: clang/lib/Driver/ToolChains/Darwin.cpp
Comment (vitalybuka) [inline clang/lib/Driver/ToolChains/Darwin.cpp:1512]: Driver needs tests

Comment (cjappl) [reply clang/lib/Driver/ToolChains/Darwin.cpp:1512]: Thanks for catching this! Added a few basic ones that test that the flag is supported, and that some known sanitizer incompatibilities are blocked

**Comment by cjappl** [comment]
Bumped the branch to resolve the conflict in config-ix.cmake

**Comment by cjappl** [comment]
Pinging reviewers @vitalybuka @MaskRay @zygoloid 

All the comments on this PR have been addressed, looking for more feedback, or approval/merge if we are getting close! Thanks in advance :)

**Comment by MaskRay** [comment]
> This is PR 1/?? (3? 4?) introducing the main guts of the realtime sanitizer. For more information, please see the [discourse thread](https://discourse.llvm.org/t/rfc-nolock-and-noalloc-attributes/76837). We have also put together a [reviewer support document](https://github.com/realtime-sanitizer/radsan/blob/doc/review-support/doc/review.md?rgh-link-date=2024-05-16T21%3A34%3A46Z) to show what our intention is.

clangDriver changes are usually the last. The expectation is that if `-fsanitize=realtime` does not return an error, there should be some basic functionality.

* clangCodeGen , testing with `%clang_cc1`
* clangDriver, testing with `%clang`

We typically split clangDriver and compiler-rt changes. Runtime compiler-rt tests often expose minor issues that only fail on certain platforms. Keeping it separate prevents us from potentially reverting the compiler-rt code, leading to unnecessary churn.
The recent numerical stability sanitizer did not do a good job for this aspect. I apologize for that as a reviewer.)

**Comment by cjappl** [comment]
> clangDriver changes are usually the last. The expectation is that if -fsanitize=realtime does not return an error, there should be some basic functionality.

Thanks for the feedback @MaskRay . I removed all of the clang components of this review, leaving only the compiler-rt components. We will re-introduce the clang changes in a future review.

In the meantime, I commented out compilation of our unit tests that required the -fsanitize flag, let me know if you'd prefer to see them removed until later. These unit tests are paired with the code introduced here, so I think at least by subject it makes sense to introduce them here, even if they are briefly not being compiled.

File: compiler-rt/lib/rtsan/tests/rtsan_test_functional.cpp
Comment (MaskRay) [inline compiler-rt/lib/rtsan/tests/rtsan_test_functional.cpp:44]: omit `{};`

File: compiler-rt/test/rtsan/lit.cfg.py
Comment (MaskRay) [inline compiler-rt/test/rtsan/lit.cfg.py:17]: compiler-rt/test/xray/lit.cfg.py uses this construct but other sanitizers don't. Is it needed?

File: compiler-rt/test/rtsan/lit.cfg.py
Comment (MaskRay) [inline compiler-rt/test/rtsan/lit.cfg.py:48]: clangdriver shall pass these arguments to the linker. lit.cfg.py should not need to do it.

File: compiler-rt/lib/rtsan/rtsan_context.cpp
Comment (MaskRay) [inline compiler-rt/lib/rtsan/rtsan_context.cpp:26]: `InternalFree`. If there is a name conflict, perhaps pick a different name.

File: compiler-rt/lib/rtsan/rtsan_context.cpp
Comment (MaskRay) [inline compiler-rt/lib/rtsan/rtsan_context.cpp:62]: https://llvm.org/docs/CodingStandards.html#use-namespace-qualifiers-to-implement-previously-declared-functions

If `using namespace __rtsan`, `.cpp` files should not need `namespace __rtsan`.

File: compiler-rt/lib/rtsan/rtsan_stack.cpp
Comment (MaskRay) [inline compiler-rt/lib/rtsan/rtsan_stack.cpp:40]: should be moved up, beside `using namespace __sanitizer;`

File: compiler-rt/lib/rtsan/rtsan_interceptors.cpp
Comment (MaskRay) [inline compiler-rt/lib/rtsan/rtsan_interceptors.cpp:19]: If other OSes are not supported, the NetBSD/FreeBSD/etc references in `lit.cfg.py` should be removed.

**Review by MaskRay — COMMENTED** [review]
Sorry for the delay. I have a really huge code review backlog...

**Comment by MaskRay** [comment]
> In the meantime, I commented out compilation of our unit tests that required the -fsanitize flag, let me know if you'd prefer to see them removed until later. These unit tests are paired with the code introduced here, so I think at least by subject it makes sense to introduce them here, even if they are briefly not being compiled.

Yeah, disabling the testsuite should be safe. The tests can be enabled after clangCodeGen and clangDriver patches landed.

**Review by MaskRay — APPROVED** [review]

File: compiler-rt/test/rtsan/lit.cfg.py
Comment (cjappl) [reply compiler-rt/test/rtsan/lit.cfg.py]: Removed!

**Comment by cjappl** [comment]
Pinging reviewers @vitalybuka @zygoloid

State of the review: 1 approval from MaskRay, all open comments have been addressed. Looking for more feedback, or more approval or a merge if this is looking good.

Thanks much! :)

File: compiler-rt/CODE_OWNERS.TXT
Comment (vitalybuka) [inline compiler-rt/CODE_OWNERS.TXT:70]: BTW @alexander-shaposhnikov NSAN is missing from here.

File: compiler-rt/lib/rtsan/tests/rtsan_test_interceptors.cpp
Comment (vitalybuka) [inline compiler-rt/lib/rtsan/tests/rtsan_test_interceptors.cpp:7]: In my experience LIT tests in are easier to maintain in LLVM environment

`TEST(TestRtsan,` are internals test, so probably fine to keep as GTEST

However `TEST(TestRtsanInterceptors` are likely better as regular LIT tests

Comment (vitalybuka) [reply compiler-rt/lib/rtsan/tests/rtsan_test_interceptors.cpp:7]: Actually even TestRtsan mostly can be done as LIT

BTW. Why I don't like GTEST in LLVM?

LIT is easier to reproduce and experiment and debug than with a particular test case of.

Also I don't trust incremental builds with GTEST, some times I am not sure that my changes are used in tests.

A couple of time CMakes bugs caused compiler-rt GTEST not executed on bots and it was unnoticed for a long time. LIT is majority, so unlikely it could happen with them.

**Review by vitalybuka — APPROVED** [review]
LGTM. but I suggest to rework test into LIT, before or either commit.

File: compiler-rt/lib/rtsan/tests/rtsan_test_interceptors.cpp
Comment (cjappl) [reply compiler-rt/lib/rtsan/tests/rtsan_test_interceptors.cpp:7]: Seems reasonable to convert these to lit! I propose leaving this for this commit, and we convert them in the near future. Specifically this next commit will introduce the -fsanitize=realtime flag, and come with a bunch of the lit tests that we have written, so it will be easier to convert them once that infrastructure has landed.

One clarifying question, for when that work is done:

Should we have a lit test for every intereceptor? The reason we did it in gtest unit tests is that adding a new interceptor's test is a few lines, as opposed to a  brand new file with the lit boilerplate. This makes it easy to make sure that every interceptor does what we think it does. We worried about the "cost" of the new file for every interceptor.  In our demo branch, as we have it laid out, the lit tests test on the full end-to-end system level, while these tests actually hit every interceptor.

**Comment by cjappl** [comment]
Also, just as a heads up, neither of the authors have commit access as we are new to LLVM. If this is OK to go in we will need help with that step.

File: compiler-rt/lib/rtsan/tests/rtsan_test_interceptors.cpp
Comment (vitalybuka) [reply compiler-rt/lib/rtsan/tests/rtsan_test_interceptors.cpp:7]: > Should we have a lit test for every intereceptor? 

I don't have strong preference. For other sanitizer default interceptor behavior is to crash, so we usually create a test per interceptor.

> The reason we did it in gtest unit tests is that adding a new interceptor's test is a few lines, as opposed to a brand new file with the lit boilerplate. This makes it easy to make sure that every interceptor does what we think it does. We worried about the "cost" of the new file for every interceptor. In our demo branch, as we have it laid out, the lit tests test on the full end-to-end system level, while these tests actually hit every interceptor.

Yes, boilerplate of LIT is not nice. So there is a trade off. I guess you can wait and see.

**Comment by vitalybuka** [comment]
I can reproduce and will debug in an hour.

**Comment by cjappl** [comment]
> I can reproduce and will debug in an hour.

Thanks for the help @vitalybuka , I am looking into this as well, but very grateful for more experienced eyes on the matter too. I'll keep you posted here if I find anything

**Comment by cjappl** [comment]
There are 3 errors I see:

Mostly it's this `llvm_gtest` issue:
```
ninja: error: 'compiler-rt/lib/rtsan/tests/llvm_gtest', needed by 'compiler-rt/lib/rtsan/tests/RtsanNoInstTestObjects.rtsan_preinit.cpp.x86_64.o', missing and no known rule to make it
```

On fuchsia it is a parameter type issue:
```
/var/lib/buildbot/fuchsia-x86_64-linux/llvm-project/compiler-rt/lib/rtsan/rtsan_interceptors.cpp:55:22: error: use of undeclared identifier '__unsanitized_open'; did you mean '__unsanitized_memset'?
   55 |   const int result = REAL(open)(path, oflag, mode);
      |                      ^~~~~~~~~~
      |                      __unsanitized_memset
/var/lib/buildbot/fuchsia-x86_64-linux/llvm-project/compiler-rt/lib/rtsan/../interception/interception.h:241:18: note: expanded from macro 'REAL'
  241 | # define REAL(x) __unsanitized_##x
      |                  ^
<scratch space>:138:1: note: expanded from here
  138 | __unsanitized_open
      | ^
/usr/local/fuchsia/sdk/arch/x64/sysroot/include/zircon/sanitizer.h:34:18: note: '__unsanitized_memset' declared here
   34 | __typeof(memset) __unsanitized_memset;
      |                  ^
/var/lib/buildbot/fuchsia-x86_64-linux/llvm-project/compiler-rt/lib/rtsan/rtsan_interceptors.cpp:55:33: error: cannot initialize a parameter of type 'void *' with an lvalue of type 'const char *'
   55 |   const int result = REAL(open)(path, oflag, mode);
      |                                 ^~~~
/usr/local/fuchsia/sdk/arch/x64/sysroot/include/zircon/sanitizer.h:34:18: note: passing argument to parameter here
   34 | __typeof(memset) __unsanitized_memset;
      |                  ^
/var/lib/buildbot/fuchsia-x86_64-linux/llvm-project/compiler-rt/lib/rtsan/rtsan_interceptors.cpp:69:22: error: use of undeclared identifier '__unsanitized_openat'
   69 |   const int result = REAL(openat)(fd, path, oflag, mode);
      |                      ^
/var/lib/buildbot/fuchsia-x86_64-linux/llvm-project/compiler-rt/lib/rtsan/../interception/interception.h:241:18: note: expanded from macro 'REAL'
  241 | # define REAL(x) __unsanitized_##x
      |                  ^
<scratch space>:141:1: note: expanded from here
  141 | __unsanitized_openat
      | ^
/var/lib/buildbot/fuchsia-x86_64-linux/llvm-project/compiler-rt/lib/rtsan/rtsan_interceptors.cpp:77:22: error: use of undeclared identifier '__unsanitized_creat'
   77 |   const int result = REAL(creat)(path, mode);
      |                      ^
/var/lib/buildbot/fuchsia-x86_64-linux/llvm-project/compiler-rt/lib/rtsan/../interception/interception.h:241:18: note: expanded from macro 'REAL'
  241 | # define REAL(x) __unsanitized_##x
      |                  ^
<scratch space>:144:1: note: expanded from here
  144 | __unsanitized_creat
      | ^
Step 6 (build) failure: build (failure)
...
```

On windows (which I think we should just disable support for, for now:

```
C:\b\slave\sanitizer-windows\llvm-project\compiler-rt\lib\rtsan\rtsan_context.cpp(19): fatal error C1083: Cannot open include file: 'pthread.h': No such file or directory
```


Working to see if I can fix the llvm_gtest issue right now. I would love advice on Fuchsia and disabling rtsan for windows.

**Comment by cjappl** [comment]
This patch I have a strong hunch fixes issue 1 in my comment above. This patch compiles and runs on my linux container and my mac, proving it doesn't affect compilation on those two systems.

```
diff --git a/compiler-rt/lib/rtsan/tests/CMakeLists.txt b/compiler-rt/lib/rtsan/tests/CMakeLists.txt
index d96e538b255f..e539ebc387e0 100644
--- a/compiler-rt/lib/rtsan/tests/CMakeLists.txt
+++ b/compiler-rt/lib/rtsan/tests/CMakeLists.txt
@@ -60,7 +60,7 @@ foreach(arch ${RTSAN_TEST_ARCH})
   #  RtsanUnitTests "Rtsan-${arch}-Test" ${arch}
   #  COMPILE_DEPS ${RTSAN_UNITTEST_HEADERS}
   #  SOURCES ${RTSAN_INST_TEST_SOURCES} ${COMPILER_RT_GOOGLETEST_SOURCES}
-  #  DEPS llvm_gtest rtsan
+  #  DEPS rtsan
   #  CFLAGS ${RTSAN_UNITTEST_CFLAGS} -fsanitize=realtime
   #  LINK_FLAGS ${RTSAN_UNITTEST_LINK_FLAGS} -fsanitize=realtime)

@@ -94,7 +94,6 @@ foreach(arch ${RTSAN_TEST_ARCH})
     COMPILE_DEPS ${RTSAN_UNITTEST_HEADERS}
     SOURCES ${RTSAN_NOINST_TEST_SOURCES}
             ${COMPILER_RT_GOOGLETEST_SOURCES}
-    DEPS llvm_gtest
     CFLAGS ${RTSAN_UNITTEST_CFLAGS}
     LINK_FLAGS ${RTSAN_UNITTEST_LINK_FLAGS}
     RUNTIME ${RTSAN_TEST_RUNTIME})
```

This also better matches the other sanitizers, llvm_gtest is a dependency in llvm, anything in compiler-rt using gtest just uses the sources directly as seen in:

```
SOURCES ${RTSAN_NOINST_TEST_SOURCES}
             ${COMPILER_RT_GOOGLETEST_SOURCES}
```

**Comment by glandium** [comment]
This also breaks building the compiler-rt for older android versions, because pthread_spinlock_t is only defined when __ANDROID_API__ >= 24.

**Comment by cjappl** [comment]
> This also breaks building the compiler-rt for older android versions, because pthread_spinlock_t is only defined when `__ANDROID_API__` >= 24.

Thanks for reporting this @glandium ! I am in the process of disabling this for android as we speak. Let me know if you see anything else fishy

**Pull request merged**
